File: scripts/ImageSaver.py
#!/usr/bin/python
import os
import rospy
from openhab_msgs.msg import ImageState
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError

# OpenCV
import cv2

VERBOSE = False

class ImageSaver(object):

    def __init__(self, item_name):
        self.item_name = item_name

        """Configure subscriber."""
        # Create a subscriber with appropriate topic, custom message and name of
        # callback function.
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", ImageState, self.callback)

        # Initialize message variables.
        self.enable = False
        self.data = None
        self.bridge = CvBridge()
        self.image = None
        self.i = 0
        self.name = "%s_%s_%s.jpg" % (self.item_name, rospy.Time.now(), self.i)

        if self.enable:
            self.start()
        else:
            self.stop()

    def start(self):
        self.enable = True
        self.sub = rospy.Subscriber("/openhab/items/" + self.item_name + "/state", ImageState, self.callback)

    def callback(self, data):
        """Handle subscriber data."""

        self.data = data

        if self.data.isnull == False:
            msg = "Received %s" % self.data.item

            try:
                self.image = self.bridge.imgmsg_to_cv2(self.data.state, 'bgr8')
            except CvBridgeError as e:
                print(e)

            self.i = self.i + 1
            cv2.imwrite(self.name, self.image)
        else:
            msg = "Received %s with NULL" % self.data.item
        rospy.loginfo(rospy.get_caller_id() + msg)

    def stop(self):
        """Turn off subscriber."""
        self.enable = False
        self.sub.unregister()

if __name__ == '__main__':
    # Initialize the node and name it.
    node_name = "ImageSaverNode"
    rospy.init_node(node_name, anonymous=True)

    imageSaver = ImageSaver("testImage")

    # Go to the main loop
    try:
        imageSaver.start()
        # Wait for messages on topic, go to callback function when new messages arrive.
        # spin() simply keeps python from exiting until this node is stopped
        rospy.spin()
    # Stop with Ctrl + C
    except KeyboardInterrupt:
        imageSaver.stop()

        nodes = os.popen("rosnode list").readlines()
        for i in range(len(nodes)):
            nodes[i] = nodes[i].replace("\n","")

        for node in nodes:
            os.system("rosnode kill " + node_name)


        print("Node stopped")

File: src/ImageSaver.cpp
#include <ros/ros.h>
#include <iostream>
#include <sstream>
#include <opencv2/highgui/highgui.hpp>
#include <cv_bridge/cv_bridge.h>
#include <sensor_msgs/Image.h>
#include "openhab_msgs/ImageState.h"

bool VERBOSE = false;

class ImageSaver
{
public:
  ImageSaver(std::string item_name)
    : item_name_(item_name)
    , enable_(false)
    , data_(nullptr)
    , i_(0)
    , nh_("~")
  {
    // Create a subscriber with appropriate topic, custom message and name of callback function.
    sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 1, &ImageSaver::callback, this);
    name_ = item_name_ + "_" + ros::Time::now().toSec() + "_" + std::to_string(i_) + ".jpg";

    if (enable_)
    {
      start();
    }
    else
    {
      stop();
    }
  }

  void start()
  {
    enable_ = true;
    sub_ = nh_.subscribe("/openhab/items/" + item_name_ + "/state", 1, &ImageSaver::callback, this);
  }

  void callback(const openhab_msgs::ImageState::ConstPtr& msg)
  {
    // Simply print out values in our custom message.
    data_ = msg;

    if (data_->isnull == false)
    {
      std::stringstream ss;
      ss << "Received " << data_->item;

      try
      {
        image_ = cv_bridge::toCvCopy(data_->state, sensor_msgs::image_encodings::BGR8)->image;
      }
      catch (cv_bridge::Exception& e)
      {
        ROS_ERROR("cv_bridge exception: %s", e.what());
        return;
      }

      i_++;
      cv::imwrite(name_, image_);
    }
    else
    {
      std::stringstream ss;
      ss << "Received " << data_->item << " with NULL";
    }
    ROS_INFO_STREAM(ros::this_node::getName() << ": " << ss.str());
  }

  void stop()
  {
    // Turn off subscriber.
    enable_ = false;
    sub_.shutdown();
  }

private:
  std::string item_name_;
  bool enable_;
  openhab_msgs::ImageState::ConstPtr data_;
  cv::Mat image_;
  int i_;
  std::string name_;
  ros::NodeHandle nh_;
  ros::Subscriber sub_;
};

int main(int argc, char** argv)
{
  // Initialize the node and name it.
  std::string node_name = "ImageSaverNode";
  ros::init(argc, argv, node_name);
  ros::NodeHandle nh;

  ImageSaver imageSaver("testImage");

  // Wait for messages on topic, go to callback function when new messages arrive.
  // spin() simply keeps cpp from exiting until this node is stopped
  try
  {
    imageSaver.start();
    ros::spin();
  }
  catch (...)
  {
    imageSaver.stop();
  }

  return 0;
}
